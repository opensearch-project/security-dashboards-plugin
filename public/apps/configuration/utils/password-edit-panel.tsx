/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import { CoreStart } from 'opensearch-dashboards/public';
import { EuiFieldText, EuiFlexGroup, EuiFlexItem, EuiFormRow, EuiIcon } from '@elastic/eui';
import { FormRow } from './form-row';
import { PASSWORD_INSTRUCTION } from '../../apps-constants';
import { getDashboardsInfo } from '../../../utils/dashboards-info-utils';
import { PasswordStrengthBar } from './password-strength-bar';

export function PasswordEditPanel(props: {
  coreStart: CoreStart;
  updatePassword: (p: string) => void;
  updateIsInvalid: (v: boolean) => void;
}) {
  const [password, setPassword] = React.useState<string>('');
  const [repeatPassword, setRepeatPassword] = React.useState<string>('');
  const [isRepeatPasswordInvalid, setIsRepeatPasswordInvalid] = React.useState<boolean>(false);
  const [passwordHelpText, setPasswordHelpText] = React.useState<string>(PASSWORD_INSTRUCTION);

  React.useEffect(() => {
    const fetchData = async () => {
      try {
        setPasswordHelpText(
          (await getDashboardsInfo(props.coreStart.http)).password_validation_error_message
        );
      } catch (e) {
        console.error(e);
      }
    };

    fetchData();
  }, [props.coreStart.http]);

  React.useEffect(() => {
    props.updatePassword(password);
    const isInvalid = repeatPassword !== password;
    setIsRepeatPasswordInvalid(isInvalid);
    props.updateIsInvalid(isInvalid);
  }, [password, props, repeatPassword]);

  const passwordChangeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
    setPassword(e.target.value);
  };

  const repeatPasswordChangeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
    setRepeatPassword(e.target.value);
  };

  return (
    <>
      <EuiFlexGroup direction="row">
        <EuiFlexItem grow={false}>
          <FormRow headerText="Password" helpText={passwordHelpText}>
            <EuiFieldText
              data-test-subj="password"
              prepend={<EuiIcon type="lock" />}
              type="password"
              onChange={passwordChangeHandler}
            />
          </FormRow>
          <EuiFormRow>
            <PasswordStrengthBar password={password} />
          </EuiFormRow>
        </EuiFlexItem>
      </EuiFlexGroup>

      <FormRow
        headerText="Re-enter password"
        helpText="The password must be identical to what you entered above."
      >
        <EuiFieldText
          data-test-subj="re-enter-password"
          prepend={<EuiIcon type="lock" />}
          type="password"
          isInvalid={isRepeatPasswordInvalid}
          onChange={repeatPasswordChangeHandler}
        />
      </FormRow>
    </>
  );
}
