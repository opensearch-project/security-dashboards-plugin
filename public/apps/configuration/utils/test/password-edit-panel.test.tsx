/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { mount, shallow } from 'enzyme';
import React from 'react';
import { PasswordEditPanel } from '../password-edit-panel';

const mockDashboardsInfo = {
  multitenancy_enabled: true,
  private_tenant_enabled: true,
  default_tenant: '',
  password_validation_error_message:
    'Password must be minimum 5 characters long and must contain at least one uppercase letter, one lowercase letter, one digit, and one special character.',
};

jest.mock('../../../../utils/dashboards-info-utils', () => ({
  getDashboardsInfo: jest.fn().mockImplementation(() => {
    return mockDashboardsInfo;
  }),
}));

describe('Password edit panel', () => {
  let component;
  const setState = jest.fn();
  const updatePassword = jest.fn();
  const updateIsInvalid = jest.fn();
  const useState = jest.spyOn(React, 'useState');
  const useEffect = jest.spyOn(React, 'useEffect');
  const mockCoreStart = {
    http: 1,
  };

  beforeEach(() => {
    useEffect.mockImplementationOnce((f) => f());
    useState.mockImplementation((initialValue) => [initialValue, setState]);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders', (done) => {
    mount(
      <PasswordEditPanel
        coreStart={mockCoreStart as any}
        updatePassword={updatePassword}
        updateIsInvalid={updateIsInvalid}
      />
    );
    process.nextTick(() => {
      expect(updatePassword).toHaveBeenCalledTimes(1);
      expect(updateIsInvalid).toHaveBeenCalledTimes(1);
      expect(setState).toBeCalledWith(mockDashboardsInfo.password_validation_error_message);
      done();
    });
  });

  it('password field update', () => {
    component = shallow(
      <PasswordEditPanel
        coreStart={mockCoreStart as any}
        updatePassword={updatePassword}
        updateIsInvalid={updateIsInvalid}
      />
    );
    const event = {
      target: { value: 'dummy' },
    } as React.ChangeEvent<HTMLInputElement>;
    component.find('[data-test-subj="password"]').simulate('change', event);
    expect(setState).toBeCalledWith('dummy');
  });

  it('repeat password field update', () => {
    component = shallow(
      <PasswordEditPanel
        coreStart={mockCoreStart as any}
        updatePassword={updatePassword}
        updateIsInvalid={updateIsInvalid}
      />
    );
    expect(component).toMatchSnapshot();
    const event = {
      target: { value: 'dummy' },
    } as React.ChangeEvent<HTMLInputElement>;
    component.find('[data-test-subj="re-enter-password"]').simulate('change', event);
    expect(setState).toBeCalledWith('dummy');
  });
});
