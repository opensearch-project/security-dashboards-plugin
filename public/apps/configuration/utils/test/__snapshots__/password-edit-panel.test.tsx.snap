// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Password edit panel repeat password field update 1`] = `
<Fragment>
  <EuiFlexGroup
    direction="row"
  >
    <EuiFlexItem
      grow={false}
    >
      <FormRow
        headerText="Password"
        helpText="Password should be at least 8 characters long and contain at least one uppercase letter, one lowercase letter, one digit, and one special character."
      >
        <EuiFieldPassword
          compressed={false}
          data-test-subj="password"
          fullWidth={false}
          isLoading={false}
          onChange={[Function]}
          type="dual"
        />
      </FormRow>
      <EuiFormRow
        describedByIds={Array []}
        display="row"
        fullWidth={false}
        hasChildLabel={true}
        hasEmptyLabelSpace={false}
        labelType="label"
      >
        <PasswordStrengthBar
          password=""
        />
      </EuiFormRow>
    </EuiFlexItem>
  </EuiFlexGroup>
  <FormRow
    headerText="Re-enter password"
    helpText="The password must be identical to what you entered above."
  >
    <EuiFieldPassword
      compressed={false}
      data-test-subj="re-enter-password"
      fullWidth={false}
      isInvalid={false}
      isLoading={false}
      onChange={[Function]}
      type="dual"
    />
  </FormRow>
</Fragment>
`;
